compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -221,6 +221,103 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
         self.eq(goal.param_env, ct.ty(), ty)?;
         self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
     }
+
+    #[instrument(level = "debug", skip(self))]
+    fn compute_uninhabited_goal(
+        &mut self,
+        Goal { param_env, predicate: (ty, module) }: Goal<'tcx, (Ty<'tcx>, Option<DefId>)>,
+    ) -> QueryResult<'tcx> {
+        let tcx = self.tcx();
+        let Some(ty) = self.try_normalize_ty(param_env, ty)? else {
+            return self.evaluate_added_goals_and_make_canonical_response(Certainty::OVERFLOW);
+        };
+        match ty.kind() {
+            ty::Never => self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes),
```

Comment:
Needs to return ambiguity on `ty::Infer(ty::TyVar(_))` -- should also ICE on fresh tys, etc.